#### Name: Therrin Baker

# PHYS 230 Lab Assignment 7

### Wednesday, February 11, 2026: Chapter 4.1-4.3
- Speed and Accuracy

#### Start by importing any packages you need below (feel free to update as you go):

In [32]:
import numpy as np
import math as m
import time

## Short python snippets (10 pts - 5 pts each)
Below each markdow cell, write a short python snippet to perform the indicated action; you will likely have to use google to find out how to perform some of these actions--this is deliberate on my part. Google and *stack overflow* are your friends.<br>
Add a markdown cell to explain your thinking if this is needed.

**1.** There are 3 ways to change the decimal accuracy that you use:
    - Using “%”:- “%” operator is used to format as well as set precision in python. 
    - Using format():- This is yet another way to format the string for setting precision.
    - Using round(x,n):- This function takes 2 arguments, number and a number of decimal points wanted. 
    
   Look up how to use these three different formatting options, and use all three to print out the number $\pi$ to 3 decimal places. 

In [29]:
i = np.e
"%.3f" %i
"{pi}".format(pi=i)


print("%.3f" %i)
print("{:.4f}".format(i))
print(round(i,5))

2.718
2.7183
2.71828


**2.** I have given the book example 4.2 below for The Madelung Constant is a dimnensionless quantity in condensed matter physics giving the total electric potential felt by an atom in a solid.  
- Edit the code so you can measure the time it takes to do the whole calculation. Then, time how long it takes the code to run 100 atoms, 1000 atoms, and 1 million. ($L$ is the number of atoms) 
    - *if you are using VScode:* yes I realize that the cells can time the runtime for you, please use code to time it as well
- Using this information, estimate how long it would take for the program to run for 1 billion terms & justify your value using markdown [do not run the code for 1 billion atoms]. 
- Also comment on the difference in the result as the number of terms increases. The expected value is around -1.74699. 

In [38]:
start = time.time()

L = 200 # number of atoms in the solid 
M = 0.0 # Medelung constant starting value 

# loop over all directions in the solid 
for i in range(-L,L+1):
    for j in range(-L,L+1):
        for k in range(-L,L+1):
            if i==0 and j==0 and k==0: #condition for if at origin 
                continue
            elif (i+j+k)%2==0: # condition for if i*j*k position is positive 
                M += 1/np.sqrt(i*i+j*j+k*k)
            else: # condition for if i*j*k position is negative 
                M -= 1/np.sqrt(i*i+j*j+k*k)

print("M =",M)
end = time.time()
print(end-start)

M = -1.7446850421707383
154.85640144348145


10 atoms: 0.024952411651611328 seconds

50 atoms: 2.4935109615325928 seconds 

100 atoms: 19.35565733909607 seconds

200 atoms: 154.85640144348145 seconds

x atoms : ~0.0000193572 * x^3 seconds, using a best fit on desmos

For 1 billion terms the code would approximately take an approximate $6\cdot 10^{23}$ years.

## Write a script 1 (25 pts)
#### Calculating a factorial  (adapted from textbook)

Write a code to do the following: 
- Write a function `factorial_int` to calculate and print the factorial of a number entered by the user and return as an interger value (see section 2.6)
- Write a second function `factorial_float` to calculate the factorial of the entered number as a floating variable instead. 
- Use each to calculate the factorial of 200 entered by a user. 
- What do you find? Include an explanation of the difference between the two in your markdown cell. 

When your program is complete, add a markdown cell below it to explain your process and program and include any values/comparisons. 

In [ ]:
def factorial_int(n):
    n = int(n)
    output = 1
    for i in range(1,n+1):
        output *=i
    return output

def factorial_float(f):
    f = int(f)
    output = 1.
    for i in range(1,f+1):
        output *=i
    return output

Input_200 = input()
Float_fact = factorial_float(Input_200)

Int_fact = factorial_int(Input_200)


print(Float_fact)
print(Int_fact)

inf
788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000


I define the functions as for loops, making them different by having internal value of the functions be defined as a float and integer respectively. I then get infinite and ~$7\cdot 10^{376}$ for the output of each of the functions.

Calculating with a float is faster, despite the code even having an additional formatting operator.

## Write a script 2 (22 pts)
####  Calculating a derivative (adapted from textbook)

Using the definition of a derivative below we can use a computer take the derivative of a function $f(x)$ at a point $x$:
$$ \frac{df}{dx} = \lim_{h \to 0} \frac{f(x+h)-f(x)}{h}$$
Actually taking $h$ to zero is not reasonable, but we can at least make $h$ small.


Write a code to do the following:
- Define a function `f1(x)` that returns the value of $x(x-1)$ [this should only take you 2 lines]
- Define a second function `f_deriv(x, h)` that takes in an x and h value and uses your `f1(x)` function to calculate the derivative of the function at the point $x = 1$ using the formula above and $h = 10^{-2}$. Have your function return both the result of the derivative and h (in that order). 
- Put the result in a markdown cell and comment on the accuracy by calculating the actual derivative at $x=1$ yourself (you can do this on your own, I know you can!). 
- Repeat the calculation for $h = 10^{-4}, 10^{-6}, 10^{-8}, 10^{-10}, 10^{-12}, 10^{-14}$. You should see an increase in accuracy and then a decrease. In the markdown cell, include all the results and comment on why this occurs. 

When your program is complete, add a markdown cell below it to explain your process and program and include any values/comparisons. 

In [ ]:
def f1(x):
    return x*(x-1)
def f_deriv(x,h):
    return (f1(x+h)-f1(x))/h
def f_true_deriv(x):
    return 2*x-x

h_values = [1e-4,1e-6,1e-8,1e-10,1e-12,1e-14]
x_input = float(input())
print("f(x) is:",f1(x_input))
print("f'(x) is:",f_true_deriv(x_input))
for h in h_values:
    print("h:",h,"Gives an approximate slope of",f_deriv(x_input,h))


f(x) is: 0.0
f'(x) is: 1.0
h: 0.0001 Gives an approximate slope of 1.0000999999998899
h: 1e-06 Gives an approximate slope of 1.0000009999177333
h: 1e-08 Gives an approximate slope of 1.0000000039225287
h: 1e-10 Gives an approximate slope of 1.000000082840371
h: 1e-12 Gives an approximate slope of 1.0000889005833413
h: 1e-14 Gives an approximate slope of 0.9992007221626509


First we define three user defined functions: the original function, the approximate derivative, and the true derivative. Lastly I define a list of h values before evaluating the approximate derivative function for each of the h values. I then print everything.

## Write a script 3 (22 pts)
####  Calculating integrals (adapted from textbook)

Like you did with a derivative, we can also use the definition of an integral to evaluate integrals with a computer. If we want to evaluate the integral 
 

$$ I = \int_{-1}^{1} \sqrt{1-x^2} \,dx, $$


we can divide the domain of the integration in to a larger number $N$ of slices of width $h = 2/N$ each. The Riemann definition can then be applied: 
$$ I = \lim_{N \to\infty} \sum_{k=1}^{N} hy_k, $$
where 
$$ y_k = \sqrt{1-x_k^2} \hspace{0.2in} \text{and}\hspace{0.2in} x_k = -1+hk.$$


We just need to make $N$ large to get a good approximation (since we can't in practice take the limit as $N\rightarrow \infty$). The expected value of this integral is $\pi/2$. 

Write a code to do the following: 
- Define a function `f2(x)` that returns the value of $\sqrt{1-x^2}.$  [this should only take you 2 lines]
- Define a second function `f_int(N)` that takes in in an N value and uses your `f2(x)` function to evaluate the integral. Have your function return both the result of the derivative and N (in that order).
- $N = 100$ and compare with the expected result. [perhaps a percent error makes a good comparison!]
- Increase the value of $N$ to make it more accurate. If the program must run in 1 second or less, how accurate can you get? 


When your program is complete, add a markdown cell below it to explain your process and program and include any comparisions. 

In [118]:

start = time.time()
def f2(x):
    return np.sqrt(1-x**2)
def f_int(N):
    Sum = 0
    delta = 2/N
    x_k = -1
    for k in range(N):
        x_k += delta
        y_k = f2(x_k)
        Sum += delta*y_k
    return Sum
Input_N = 1000000
print(f_int(Input_N))
end = time.time()
print(end-start, "seconds")
print(np.pi/2)
print(np.abs(f_int(Input_N)-np.pi/2))

1.5707963251616452
0.7659759521484375 seconds
1.5707963267948966
1.633251311972117e-09


For the code above I start by defining the start time of the code, and then define two user defined functions. I can get the value to within 1e-9 of $\pi/2$ with $N=1000000$. 